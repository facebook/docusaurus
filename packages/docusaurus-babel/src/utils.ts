/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import {BABEL_CONFIG_FILE_NAME} from '@docusaurus/utils';
import type {TransformOptions} from '@babel/core';

export async function getCustomBabelConfigFilePath(
  siteDir: string,
): Promise<string | undefined> {
  const customBabelConfigurationPath = path.join(
    siteDir,
    BABEL_CONFIG_FILE_NAME,
  );
  return (await fs.pathExists(customBabelConfigurationPath))
    ? customBabelConfigurationPath
    : undefined;
}

export function getBabelOptions({
  isServer,
  babelOptions,
}: {
  isServer?: boolean;
  // TODO Docusaurus v4 fix this
  //  weird to have getBabelOptions take a babelOptions param
  babelOptions?: TransformOptions | string;
} = {}): TransformOptions {
  const caller = {name: isServer ? 'server' : 'client'};
  if (typeof babelOptions === 'string') {
    return {
      babelrc: false,
      configFile: babelOptions,
      caller,
    };
  }
  return {
    ...(babelOptions ?? {
      presets: [require.resolve('@docusaurus/babel/preset')],
    }),
    babelrc: false,
    configFile: false,
    caller,
  };
}
