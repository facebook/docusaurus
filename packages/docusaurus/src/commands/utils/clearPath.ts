/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {rm} from 'fs/promises';
import {PerfLogger} from '@docusaurus/logger';

/**
 * @param pathToClear
 */
export default async function clearPath(pathToClear: string): Promise<void> {
  return PerfLogger.async(
    `clearPath ${path.relative(process.cwd(), pathToClear)}`,
    async () => {
      await rm(pathToClear, {recursive: true, force: true});
    },
  );
}
