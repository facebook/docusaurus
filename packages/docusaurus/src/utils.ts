/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import {AsyncLocalStorage} from 'async_hooks';
import logger from '@docusaurus/logger';

// For now this is a private env variable we use internally
// But we'll want to expose this feature officially some day
export const PerfDebuggingEnabled: boolean =
  !!process.env.DOCUSAURUS_PERF_LOGGER;

const Thresholds = {
  min: 5,
  yellow: 100,
  red: 1000,
};

const PerfPrefix = logger.yellow(`[PERF] `);

// This is what enables to "see the parent stack" for each log
// Parent1 > Parent2 > Parent3 > child trace
const ParentPrefix = new AsyncLocalStorage<string>();
function applyParentPrefix(label: string) {
  const parentPrefix = ParentPrefix.getStore();
  return parentPrefix ? `${parentPrefix} > ${label}` : label;
}

type PerfLoggerAPI = {
  start: (label: string) => void;
  end: (label: string) => void;
  log: (message: string) => void;
  async: <Result>(
    label: string,
    asyncFn: () => Result | Promise<Result>,
  ) => Promise<Result>;
};

function createPerfLogger(): PerfLoggerAPI {
  if (!PerfDebuggingEnabled) {
    const noop = () => {};
    return {
      start: noop,
      end: noop,
      log: noop,
      async: async (_label, asyncFn) => asyncFn(),
    };
  }

  const formatDuration = (duration: number): string => {
    if (duration > Thresholds.red) {
      return logger.red(`${(duration / 1000).toFixed(2)} seconds!`);
    } else if (duration > Thresholds.yellow) {
      return logger.yellow(`${duration.toFixed(2)} ms`);
    } else {
      return logger.green(`${duration.toFixed(2)} ms`);
    }
  };

  const logDuration = (label: string, duration: number) => {
    if (duration < Thresholds.min) {
      return;
    }
    console.log(`${PerfPrefix + label} - ${formatDuration(duration)}`);
  };

  const start: PerfLoggerAPI['start'] = (label) => performance.mark(label);

  const end: PerfLoggerAPI['end'] = (label) => {
    const {duration} = performance.measure(label);
    performance.clearMarks(label);
    logDuration(applyParentPrefix(label), duration);
  };

  const log: PerfLoggerAPI['log'] = (label: string) =>
    console.log(PerfPrefix + applyParentPrefix(label));

  const async: PerfLoggerAPI['async'] = async (label, asyncFn) => {
    const finalLabel = applyParentPrefix(label);
    const before = performance.now();
    const result = await ParentPrefix.run(finalLabel, () => asyncFn());
    const duration = performance.now() - before;
    logDuration(finalLabel, duration);
    return result;
  };

  return {
    start,
    end,
    log,
    async,
  };
}

export const PerfLogger: PerfLoggerAPI = createPerfLogger();
