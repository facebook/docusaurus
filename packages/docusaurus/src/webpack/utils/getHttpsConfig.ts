/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import crypto from 'crypto';
import logger from '@docusaurus/logger';

// Ensure the certificate and key provided are valid and if not
// throw an easy to debug error
function validateKeyAndCerts({
  cert,
  key,
  keyFile,
  crtFile,
}: {
  cert: Buffer;
  key: Buffer;
  keyFile: string;
  crtFile: string;
}) {
  let encrypted: Buffer;
  try {
    // publicEncrypt will throw an error with an invalid cert
    encrypted = crypto.publicEncrypt(cert, Buffer.from('test'));
  } catch (err) {
    logger.error`The certificate path=${crtFile} is invalid.`;
    throw err;
  }

  try {
    // privateDecrypt will throw an error with an invalid key
    crypto.privateDecrypt(key, encrypted);
  } catch (err) {
    logger.error`The certificate key path=${keyFile} is invalid.`;
    throw err;
  }
}

// Read file and throw an error if it doesn't exist
async function readEnvFile(file: string, type: string) {
  if (!(await fs.pathExists(file))) {
    throw new Error(
      `You specified ${type} in your env, but the file "${file}" can't be found.`,
    );
  }
  return fs.readFile(file);
}

// Get the https config
// Return cert files if provided in env, otherwise just true or false
export default async function getHttpsConfig(): Promise<
  boolean | {cert: Buffer; key: Buffer}
> {
  const appDirectory = await fs.realpath(process.cwd());
  const {SSL_CRT_FILE, SSL_KEY_FILE, HTTPS} = process.env;
  const isHttps = HTTPS === 'true';

  if (isHttps && SSL_CRT_FILE && SSL_KEY_FILE) {
    const crtFile = path.resolve(appDirectory, SSL_CRT_FILE);
    const keyFile = path.resolve(appDirectory, SSL_KEY_FILE);
    const config = {
      cert: await readEnvFile(crtFile, 'SSL_CRT_FILE'),
      key: await readEnvFile(keyFile, 'SSL_KEY_FILE'),
    };

    validateKeyAndCerts({...config, keyFile, crtFile});
    return config;
  }
  return isHttps;
}
