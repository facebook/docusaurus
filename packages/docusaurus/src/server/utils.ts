/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {posixPath, Globby} from '@docusaurus/utils';

// Globby that fix Windows path patterns
// See https://github.com/facebook/docusaurus/pull/4222#issuecomment-795517329
export async function safeGlobby(
  patterns: string[],
  options?: Globby.GlobbyOptions,
): Promise<string[]> {
  // Required for Windows support, as paths using \ should not be used by globby
  // (also using the windows hard drive prefix like c: is not a good idea)
  const globPaths = patterns.map((dirPath) =>
    posixPath(path.relative(process.cwd(), dirPath)),
  );

  return Globby(globPaths, options);
}
