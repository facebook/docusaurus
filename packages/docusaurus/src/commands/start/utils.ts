/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import _ from 'lodash';
import {prepareUrls} from 'react-dev-utils/WebpackDevServerUtils';
import {normalizeUrl} from '@docusaurus/utils';
import logger from '@docusaurus/logger';
import {getHostPort} from '../../server/getHostPort';
import {PerfLogger} from '../../utils';
import {
  loadSite,
  type LoadSiteParams,
  reloadSite,
  reloadSitePlugin,
} from '../../server/site';
import {formatPluginName} from '../../server/plugins/pluginsUtils';
import type {StartCLIOptions} from './start';
import type {LoadedPlugin, RouterType} from '@docusaurus/types';

export type OpenUrlContext = {
  host: string;
  port: number;
  getOpenUrl: ({
    baseUrl,
    router,
  }: {
    baseUrl: string;
    router: RouterType;
  }) => string;
};

export async function createOpenUrlContext({
  cliOptions,
}: {
  cliOptions: StartCLIOptions;
}): Promise<OpenUrlContext> {
  const protocol: string = process.env.HTTPS === 'true' ? 'https' : 'http';

  const {host, port} = await getHostPort(cliOptions);
  if (port === null) {
    return process.exit();
  }

  const getOpenUrl: OpenUrlContext['getOpenUrl'] = ({baseUrl, router}) => {
    const urls = prepareUrls(protocol, host, port);
    return normalizeUrl([
      urls.localUrlForBrowser,
      router === 'hash' ? '/#/' : '',
      baseUrl,
    ]);
  };

  return {host, port, getOpenUrl};
}

type StartParams = {
  siteDirParam: string;
  cliOptions: Partial<StartCLIOptions>;
};

async function createLoadSiteParams({
  siteDirParam,
  cliOptions,
}: StartParams): Promise<LoadSiteParams> {
  const siteDir = await fs.realpath(siteDirParam);
  return {
    siteDir,
    config: cliOptions.config,
    locale: cliOptions.locale,
    localizePath: undefined, // Should this be configurable?
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function createReloadableSite(startParams: StartParams) {
  const openUrlContext = await createOpenUrlContext(startParams);

  const loadSiteParams = await PerfLogger.async('createLoadSiteParams', () =>
    createLoadSiteParams(startParams),
  );

  let site = await PerfLogger.async('Load site', () =>
    loadSite(loadSiteParams),
  );

  const get = () => site;

  const getOpenUrl = () =>
    openUrlContext.getOpenUrl({
      baseUrl: site.props.baseUrl,
      router: site.props.siteConfig.future.experimental_router,
    });

  const printOpenUrlMessage = () => {
    logger.success`Docusaurus website is running at: url=${getOpenUrl()}`;
  };
  printOpenUrlMessage();

  const reloadBase = async () => {
    try {
      const oldSite = site;
      site = await PerfLogger.async('Reload site', () => reloadSite(site));
      if (oldSite.props.baseUrl !== site.props.baseUrl) {
        printOpenUrlMessage();
      }
    } catch (e) {
      logger.error('Site reload failure');
      console.error(e);
    }
  };

  // TODO instead of debouncing we should rather add AbortController support?
  const reload = _.debounce(reloadBase, 500);

  // TODO this could be subject to plugin reloads race conditions
  //  In practice, it is not likely the user will hot reload 2 plugins at once
  //  but we should still support it and probably use a task queuing system
  const reloadPlugin = async (plugin: LoadedPlugin) => {
    try {
      site = await PerfLogger.async(
        `Reload site plugin ${formatPluginName(plugin)}`,
        () => {
          const pluginIdentifier = {name: plugin.name, id: plugin.options.id};
          return reloadSitePlugin(site, pluginIdentifier);
        },
      );
    } catch (e) {
      logger.error(
        `Site plugin reload failure - Plugin ${formatPluginName(plugin)}`,
      );
      console.error(e);
    }
  };

  return {get, getOpenUrl, reload, reloadPlugin, openUrlContext};
}
