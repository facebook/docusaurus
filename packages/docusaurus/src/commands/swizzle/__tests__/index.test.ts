/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {jest} from '@jest/globals';
import path from 'path';
import fs from 'fs-extra';
import tree from 'tree-node-cli';
import {escapePath, Globby, posixPath} from '@docusaurus/utils';
import {ThemePath, createTempSiteDir, Components} from './testUtils';
import {swizzle} from '../index';

const FixtureThemeName = 'fixture-theme-name';

// TODO is it really worth it to duplicate fixtures?
const ThemePathJS = ThemePath;
const ThemePathTS = ThemePath;

async function createTestSiteConfig(siteDir: string) {
  const configPath = path.join(siteDir, 'docusaurus.config.js');

  await fs.writeFile(
    configPath,
    `
module.exports = {
  title: 'My Site',
  tagline: 'Dinosaurs are cool',
  url: 'https://your-docusaurus-site.example.com',
  baseUrl: '/',
  themes: [
    function fixtureTheme() {
      return {
        name: '${FixtureThemeName}',
        getThemePath() {
          return '${escapePath(ThemePathJS)}';
        },
        getTypeScriptThemePath() {
          return '${escapePath(ThemePathTS)}';
        },
      };
    },
  ],
}`,
  );
}

class MockExitError extends Error {
  constructor(public code: number) {
    super(`Exit with code ${code}`);
    this.code = code;
  }
}

function createExitMock() {
  let mock: jest.SpyInstance<(code?: number) => never>;

  // eslint-disable-next-line jest/require-top-level-describe
  beforeEach(async () => {
    mock = jest.spyOn(process, 'exit').mockImplementation((code) => {
      throw new MockExitError(code!);
    }) as jest.SpyInstance<(code?: number) => never>;
  });
  // eslint-disable-next-line jest/require-top-level-describe
  afterEach(async () => {
    mock.mockRestore();
  });

  return {
    expectExitCode: (code: number) => {
      expect(mock).toHaveBeenCalledWith(code);
    },
  };
}

const swizzleWithExit: typeof swizzle = async (...args) => {
  await expect(() => swizzle(...args)).rejects.toThrow(MockExitError);
};

async function createTestSite() {
  const siteDir = await createTempSiteDir();

  await createTestSiteConfig(siteDir);

  const siteThemePath = path.join(siteDir, 'src/theme');
  await fs.ensureDir(siteThemePath);

  async function snapshotThemeDir() {
    const siteThemePathPosix = posixPath(siteThemePath);
    expect(tree(siteThemePathPosix)).toMatchSnapshot('theme dir tree');

    const files = (await Globby(siteThemePathPosix))
      .map((file) => path.posix.relative(siteThemePathPosix, file))
      .sort();

    for (const file of files) {
      const fileContent = await fs.readFile(
        path.posix.join(siteThemePath, file),
        'utf-8',
      );
      expect(fileContent).toMatchSnapshot(file);
    }
  }

  function testWrap({
    component,
    typescript,
  }: {
    component: string;
    typescript?: boolean;
  }) {
    return swizzleWithExit(FixtureThemeName, component, siteDir, {
      wrap: true,
      danger: true,
      typescript,
      javascript: !typescript,
    });
  }

  function testEject({
    component,
    typescript,
  }: {
    component: string;
    typescript?: boolean;
  }) {
    return swizzleWithExit(FixtureThemeName, component, siteDir, {
      eject: true,
      danger: true,
      typescript,
      javascript: !typescript,
    });
  }

  return {
    siteDir,
    siteThemePath,
    snapshotThemeDir,
    testWrap,
    testEject,
  };
}

describe('swizzle wrap', () => {
  const exitMock = createExitMock();

  it(`${Components.FirstLevelComponent} JS`, async () => {
    const {snapshotThemeDir, testWrap} = await createTestSite();
    await testWrap({
      component: Components.FirstLevelComponent,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.FirstLevelComponent} TS`, async () => {
    const {snapshotThemeDir, testWrap} = await createTestSite();
    await testWrap({
      component: Components.FirstLevelComponent,
      typescript: true,
    });
    await snapshotThemeDir();
  });

  it(`${Components.ComponentInFolder} JS`, async () => {
    const {snapshotThemeDir, testWrap} = await createTestSite();
    await testWrap({
      component: Components.ComponentInFolder,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.ComponentInFolder} TS`, async () => {
    const {snapshotThemeDir, testWrap} = await createTestSite();
    await testWrap({
      component: Components.ComponentInFolder,
      typescript: true,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.ComponentInSubFolder} JS`, async () => {
    const {snapshotThemeDir, testWrap} = await createTestSite();
    await testWrap({
      component: Components.ComponentInSubFolder,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.ComponentInSubFolder} TS`, async () => {
    const {snapshotThemeDir, testWrap} = await createTestSite();
    await testWrap({
      component: Components.ComponentInSubFolder,
      typescript: true,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.Sibling} JS`, async () => {
    const {snapshotThemeDir, testWrap} = await createTestSite();
    await testWrap({
      component: Components.Sibling,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.Sibling} TS`, async () => {
    const {snapshotThemeDir, testWrap} = await createTestSite();
    await testWrap({
      component: Components.Sibling,
      typescript: true,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });
});

describe('swizzle eject', () => {
  const exitMock = createExitMock();

  it(`${Components.FirstLevelComponent} JS`, async () => {
    const {snapshotThemeDir, testEject} = await createTestSite();
    await testEject({
      component: Components.FirstLevelComponent,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.FirstLevelComponent} TS`, async () => {
    const {snapshotThemeDir, testEject} = await createTestSite();
    await testEject({
      component: Components.FirstLevelComponent,
      typescript: true,
    });
    await snapshotThemeDir();
  });

  it(`${Components.ComponentInFolder} JS`, async () => {
    const {snapshotThemeDir, testEject} = await createTestSite();
    await testEject({
      component: Components.ComponentInFolder,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.ComponentInFolder} TS`, async () => {
    const {snapshotThemeDir, testEject} = await createTestSite();
    await testEject({
      component: Components.ComponentInFolder,
      typescript: true,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.ComponentInSubFolder} JS`, async () => {
    const {snapshotThemeDir, testEject} = await createTestSite();
    await testEject({
      component: Components.ComponentInSubFolder,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.ComponentInSubFolder} TS`, async () => {
    const {snapshotThemeDir, testEject} = await createTestSite();
    await testEject({
      component: Components.ComponentInSubFolder,
      typescript: true,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.Sibling} JS`, async () => {
    const {snapshotThemeDir, testEject} = await createTestSite();
    await testEject({
      component: Components.Sibling,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });

  it(`${Components.Sibling} TS`, async () => {
    const {snapshotThemeDir, testEject} = await createTestSite();
    await testEject({
      component: Components.Sibling,
      typescript: true,
    });
    exitMock.expectExitCode(0);
    await snapshotThemeDir();
  });
});
