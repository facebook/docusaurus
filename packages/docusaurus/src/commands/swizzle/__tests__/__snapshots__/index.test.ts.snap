// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swizzle eject ComponentInFolder JS: ComponentInFolder/ComponentInSubFolder/index.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder JS: ComponentInFolder/ComponentInSubFolder/styles.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder JS: ComponentInFolder/ComponentInSubFolder/styles.module.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder JS: ComponentInFolder/Sibling.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder JS: ComponentInFolder/index.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder JS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    ├── ComponentInSubFolder
    │   ├── index.css
    │   ├── styles.css
    │   └── styles.module.css
    ├── Sibling.css
    └── index.css"
`;

exports[`swizzle eject ComponentInFolder TS: ComponentInFolder/ComponentInSubFolder/index.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder TS: ComponentInFolder/ComponentInSubFolder/index.tsx 1`] = `
"import React from 'react';

export default function ComponentInSubFolder() {
  return <div>ComponentInSubFolder</div>;
}
"
`;

exports[`swizzle eject ComponentInFolder TS: ComponentInFolder/ComponentInSubFolder/styles.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder TS: ComponentInFolder/ComponentInSubFolder/styles.module.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder TS: ComponentInFolder/Sibling.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder TS: ComponentInFolder/Sibling.tsx 1`] = `
"import React from 'react';

export default function Sibling() {
  return <div>Sibling</div>;
}
"
`;

exports[`swizzle eject ComponentInFolder TS: ComponentInFolder/index.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder TS: ComponentInFolder/index.tsx 1`] = `
"import React from 'react';

export default function ComponentInFolder() {
  return <div>ComponentInFolder</div>;
}
"
`;

exports[`swizzle eject ComponentInFolder TS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    ├── ComponentInSubFolder
    │   ├── index.css
    │   ├── index.tsx
    │   ├── styles.css
    │   └── styles.module.css
    ├── Sibling.css
    ├── Sibling.tsx
    ├── index.css
    └── index.tsx"
`;

exports[`swizzle eject ComponentInFolder/ComponentInSubFolder JS: ComponentInFolder/ComponentInSubFolder/index.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder/ComponentInSubFolder JS: ComponentInFolder/ComponentInSubFolder/styles.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder/ComponentInSubFolder JS: ComponentInFolder/ComponentInSubFolder/styles.module.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder/ComponentInSubFolder JS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    └── ComponentInSubFolder
        ├── index.css
        ├── styles.css
        └── styles.module.css"
`;

exports[`swizzle eject ComponentInFolder/ComponentInSubFolder TS: ComponentInFolder/ComponentInSubFolder/index.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder/ComponentInSubFolder TS: ComponentInFolder/ComponentInSubFolder/index.tsx 1`] = `
"import React from 'react';

export default function ComponentInSubFolder() {
  return <div>ComponentInSubFolder</div>;
}
"
`;

exports[`swizzle eject ComponentInFolder/ComponentInSubFolder TS: ComponentInFolder/ComponentInSubFolder/styles.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder/ComponentInSubFolder TS: ComponentInFolder/ComponentInSubFolder/styles.module.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder/ComponentInSubFolder TS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    └── ComponentInSubFolder
        ├── index.css
        ├── index.tsx
        ├── styles.css
        └── styles.module.css"
`;

exports[`swizzle eject ComponentInFolder/Sibling JS: ComponentInFolder/Sibling.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder/Sibling JS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    └── Sibling.css"
`;

exports[`swizzle eject ComponentInFolder/Sibling TS: ComponentInFolder/Sibling.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject ComponentInFolder/Sibling TS: ComponentInFolder/Sibling.tsx 1`] = `
"import React from 'react';

export default function Sibling() {
  return <div>Sibling</div>;
}
"
`;

exports[`swizzle eject ComponentInFolder/Sibling TS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    ├── Sibling.css
    └── Sibling.tsx"
`;

exports[`swizzle eject FirstLevelComponent JS: FirstLevelComponent.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject FirstLevelComponent JS: theme dir tree 1`] = `
"theme
└── FirstLevelComponent.css"
`;

exports[`swizzle eject FirstLevelComponent TS: FirstLevelComponent.css 1`] = `
".testClass {
  background: black;
}
"
`;

exports[`swizzle eject FirstLevelComponent TS: FirstLevelComponent.tsx 1`] = `
"import React from 'react';

export default function FirstLevelComponent() {
  return <div>First level component</div>;
}
"
`;

exports[`swizzle eject FirstLevelComponent TS: theme dir tree 1`] = `
"theme
├── FirstLevelComponent.css
└── FirstLevelComponent.tsx"
`;

exports[`swizzle wrap ComponentInFolder JS: ComponentInFolder/index.js 1`] = `
"import React from 'react';
import ComponentInFolder from '@theme-original/ComponentInFolder';

export default function ComponentInFolderWrapper(props) {
  return (
    <>
      <ComponentInFolder {...props} />
    </>
  );
}
"
`;

exports[`swizzle wrap ComponentInFolder JS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    └── index.js"
`;

exports[`swizzle wrap ComponentInFolder TS: ComponentInFolder/index.tsx 1`] = `
"import React from 'react';
import ComponentInFolder from '@theme-original/ComponentInFolder';
import type ComponentInFolderType from '@theme/ComponentInFolder';
import type {WrapperProps} from '@docusaurus/types';

type Props = WrapperProps<typeof ComponentInFolderType>;

export default function ComponentInFolderWrapper(props: Props): JSX.Element {
  return (
    <>
      <ComponentInFolder {...props} />
    </>
  );
}
"
`;

exports[`swizzle wrap ComponentInFolder TS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    └── index.tsx"
`;

exports[`swizzle wrap ComponentInFolder/ComponentInSubFolder JS: ComponentInFolder/ComponentInSubFolder/index.js 1`] = `
"import React from 'react';
import ComponentInSubFolder from '@theme-original/ComponentInFolder/ComponentInSubFolder';

export default function ComponentInSubFolderWrapper(props) {
  return (
    <>
      <ComponentInSubFolder {...props} />
    </>
  );
}
"
`;

exports[`swizzle wrap ComponentInFolder/ComponentInSubFolder JS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    └── ComponentInSubFolder
        └── index.js"
`;

exports[`swizzle wrap ComponentInFolder/ComponentInSubFolder TS: ComponentInFolder/ComponentInSubFolder/index.tsx 1`] = `
"import React from 'react';
import ComponentInSubFolder from '@theme-original/ComponentInFolder/ComponentInSubFolder';
import type ComponentInSubFolderType from '@theme/ComponentInFolder/ComponentInSubFolder';
import type {WrapperProps} from '@docusaurus/types';

type Props = WrapperProps<typeof ComponentInSubFolderType>;

export default function ComponentInSubFolderWrapper(props: Props): JSX.Element {
  return (
    <>
      <ComponentInSubFolder {...props} />
    </>
  );
}
"
`;

exports[`swizzle wrap ComponentInFolder/ComponentInSubFolder TS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    └── ComponentInSubFolder
        └── index.tsx"
`;

exports[`swizzle wrap ComponentInFolder/Sibling JS: ComponentInFolder/Sibling.js 1`] = `
"import React from 'react';
import Sibling from '@theme-original/ComponentInFolder/Sibling';

export default function SiblingWrapper(props) {
  return (
    <>
      <Sibling {...props} />
    </>
  );
}
"
`;

exports[`swizzle wrap ComponentInFolder/Sibling JS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    └── Sibling.js"
`;

exports[`swizzle wrap ComponentInFolder/Sibling TS: ComponentInFolder/Sibling.tsx 1`] = `
"import React from 'react';
import Sibling from '@theme-original/ComponentInFolder/Sibling';
import type SiblingType from '@theme/ComponentInFolder/Sibling';
import type {WrapperProps} from '@docusaurus/types';

type Props = WrapperProps<typeof SiblingType>;

export default function SiblingWrapper(props: Props): JSX.Element {
  return (
    <>
      <Sibling {...props} />
    </>
  );
}
"
`;

exports[`swizzle wrap ComponentInFolder/Sibling TS: theme dir tree 1`] = `
"theme
└── ComponentInFolder
    └── Sibling.tsx"
`;

exports[`swizzle wrap FirstLevelComponent JS: FirstLevelComponent.js 1`] = `
"import React from 'react';
import FirstLevelComponent from '@theme-original/FirstLevelComponent';

export default function FirstLevelComponentWrapper(props) {
  return (
    <>
      <FirstLevelComponent {...props} />
    </>
  );
}
"
`;

exports[`swizzle wrap FirstLevelComponent JS: theme dir tree 1`] = `
"theme
└── FirstLevelComponent.js"
`;

exports[`swizzle wrap FirstLevelComponent TS: FirstLevelComponent.tsx 1`] = `
"import React from 'react';
import FirstLevelComponent from '@theme-original/FirstLevelComponent';
import type FirstLevelComponentType from '@theme/FirstLevelComponent';
import type {WrapperProps} from '@docusaurus/types';

type Props = WrapperProps<typeof FirstLevelComponentType>;

export default function FirstLevelComponentWrapper(props: Props): JSX.Element {
  return (
    <>
      <FirstLevelComponent {...props} />
    </>
  );
}
"
`;

exports[`swizzle wrap FirstLevelComponent TS: theme dir tree 1`] = `
"theme
└── FirstLevelComponent.tsx"
`;
